3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala

```diff
@@ -72,233 +74,288 @@ object EnrichmentManager {
+        // This enrichment cannot fail
+        setEventFingerprint(event, sourceMap, registry.eventFingerprint)
 
-    // Validate unstructured event
-    val unstructEvent
-      : F[ValidatedNel[FailureDetails.EnrichmentStageIssue, Option[SelfDescribingData[Json]]]] =
-      Shredder.extractAndValidateUnstructEvent(event, client).map(_.toValidatedNel)
+        // Execute cookie extractor enrichment
+        val cookieExtractorContexts: List[SelfDescribingData[Json]] =
+          headerContexts[CookieExtractorEnrichment](
+            raw.context.headers,
+            registry.cookieExtractor,
+            (e, hs) => e.extract(hs)
+          )
 
-    // Execute IP lookup enrichment
-    val geoloc = geoLocation(event, registry.ipLookups)
+        // Execute header extractor enrichment
+        val httpHeaderExtractorContexts: List[SelfDescribingData[Json]] =
+          headerContexts[HttpHeaderExtractorEnrichment](
+            raw.context.headers,
+            registry.httpHeaderExtractor,
+            (e, hs) => e.extract(hs)
+          )
 
-    // Execute the JavaScript scripting enrichment
-    val jsScript: Either[FailureDetails.EnrichmentStageIssue, List[SelfDescribingData[Json]]] =
-      getJsScript(event, registry.javascriptScript)
+        // Fetch weather context
+        val weatherContext: F[
+          Either[NonEmptyList[FailureDetails.EnrichmentFailure], Option[SelfDescribingData[Json]]]
+        ] =
+          getWeatherContext(event, registry.weather)
+
+        // Runs YAUAA enrichment (gets info thanks to user agent)
+        val yauaaContext: Option[SelfDescribingData[Json]] =
+          getYauaaContext(event, registry.yauaa)
+
+        // Extract the event vendor/name/format/version
+        val extractSchema: Either[FailureDetails.EnrichmentFailure, Unit] =
+          extractSchemaFields(event, unstructEvent)
+
+        // Execute IP lookup enrichment
+        val geoloc = geoLocation(event, registry.ipLookups)
+
+        // Execute the JavaScript scripting enrichment
+        val jsScript: Either[FailureDetails.EnrichmentFailure, List[SelfDescribingData[Json]]] =
+          getJsScript(event, registry.javascriptScript)
+
+        // Enrichments that don't run in a Monad
+        val noMonadEnrichments: ValidatedNel[FailureDetails.EnrichmentFailure, Unit] =
+          (
+            collectorTstamp.toValidatedNel,
+            useragent.toValidatedNel,
+            collectorVersionSet.toValidatedNel,
+            pageUri.toValidatedNel,
+            derivedTstamp.toValidatedNel,
+            iabContext.toValidated,
+            uaUtils.toValidatedNel,
+            uaParser.toValidatedNel,
+            refererUri.toValidatedNel,
+            pageQsMap.toValidatedNel,
+            crossDomain.toValidatedNel,
+            extractSchema.toValidatedNel,
+            jsScript.toValidatedNel
+          ).mapN((_, _, _, _, _, _, _, _, _, _, _, _, _) => ())
```

Comment:
I just realized that it can be:

```suggestion
          List(
            collectorTstamp.toValidatedNel,
            useragent.toValidatedNel,
            collectorVersionSet.toValidatedNel,
            pageUri.toValidatedNel,
            derivedTstamp.toValidatedNel,
            iabContext.toValidated,
            uaUtils.toValidatedNel,
            uaParser.toValidatedNel,
            refererUri.toValidatedNel,
            pageQsMap.toValidatedNel,
            crossDomain.toValidatedNel,
            extractSchema.toValidatedNel,
            jsScript.toValidatedNel
          ).sequence_
```

So we don't need to care about arity.